src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,11 +124,68 @@
       (.setAs (clojure-response-body-type->java opts))
       (.setBody (:body opts))
       (.setDecompressBody (clojure.core/get opts :decompress-body true))
-      (.setHeaders (:headers opts))))
+      (.setHeaders (:headers opts))
+      (.setMetricId (parse-metric-id opts))))
+
+(schema/defn get-mean :- schema/Num
+  [timer :- Timer]
+  (->> timer
+       .getSnapshot
+       .getMean
+       (.toMillis TimeUnit/NANOSECONDS)))
+
+(defn get-metric-data
+  [timer metric-id]
+  (let [count (.getCount timer)
+        mean (get-mean timer)
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate
+     :metric-id metric-id}))
+
+(defn get-metrics-data
+  [timers]
+  (reduce (fn [acc [metric-id timer]]
+            (assoc acc metric-id (get-metric-data timer metric-id)))
+          {} timers))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (when metric-registry
+     (into {} (Metrics/getClientMetrics metric-registry))))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
+   (when metric-registry
+     (cond
+       (:verb metric-filter) (into {} (Metrics/getClientMetricsWithUrlAndVerb
+                                       metric-registry
+                                       (:url metric-filter)
+                                       (:verb metric-filter)
+                                       Metrics$MetricType/BYTES_READ))
+       (:url metric-filter) (into {} (Metrics/getClientMetricsWithUrl
+                                      metric-registry
+                                      (:url metric-filter)
+                                      Metrics$MetricType/BYTES_READ))
+       (:metric-id metric-filter) (into {} (Metrics/getClientMetricsWithMetricId
+                                            metric-registry
+                                            (into-array (:metric-id metric-filter)) Metrics$MetricType/BYTES_READ))
+       :else (throw (IllegalArgumentException. "Not an allowed metric filter."))))))
+
+(schema/defn ^:always-validate get-client-metrics-data :- (schema/maybe common/MetricsData)
+  "Returns a map of metric-id to metric data summary."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
```

Comment:
I wasn't sure exactly the best place to put some general API feedback, but I hate how github won't let you "reply" to a comment that isn't attached to a line of code, and I think most of this comes back to the schema for the `MetricFilter`, so this seemed like a reasonable place.  Rolling up a few things here:

1. I think we may want to monkey with some of the names, e.g. `with-metric-id`, `bytes-read`, etc., but I feel like it should be pretty easy to put that off as like the last thing we do in this PR (or maybe even a follow-on PR).  So I'd vote for maybe belaying that conversation.
2. Constructing one of these `MetricFilter` schema objects by hand feels a bit cumbersome (only talking about Clojure here, I don't necessarily care whether the Java API is a bit cumbersome, because, you know, it's Java).  I think it might be worth creating some helper functions to help build these up.  Maybe, at the least, a function that you pass in a partial map and it fills in the rest with some default values?  Or, to get a bit more granular, maybe something like `(metric-id-filter ["puppetlabs"])` and `(url-filter "https://tickets.puppetlabs.com")`.
3. I notice that if I make a request to `https://confluence.puppetlabs.com` and then another one to `https://confluence.puppetlabs.com/` (with a trailing slash), these show up as two separate metrics on the `with-url` side of the house.  I think that's probably fine, and that if people wanted to aggregate those somehow they can always use a `metric-id`.  Just thought it was worth mentioning; it also made me wonder if we needed to think at all about special characters that might show up in a URL and whether there are any possible issues with how those might get mapped to a metric id.
4. Assuming we allow raw URLs (including funky chars), to show up in the `with-url` metric ids, we should probably think a little about what might happen if someone exports these metrics to graphite.  I've found in the past that because of general issues around character escaping, it can be challenging to use a GUI like grafana to render a metric if the metric id has special characters in it, because in grafana you might have to do something like prepend the characters with backslashes or URL-encode them in order for it to find them in graphite.  I *think* it's probably OK for us to say that people should only consider exporting the `with-metric-id` metrics to graphite, and that that would probably alleviate this issue, but just something worth thinking about and potentially codifying in some docs or something.
5. I notice that query strings are included in the `with-url` metric ids.  Mixed feelings on this; I can see how they might be useful, but I can also imagine that leading to a large proliferation of timers, and make it hard or impossible to aggregate the URL metrics for a given URL ignoring the differences in the query string.  So, I could imagine us making the call that we just truncate query strings from the `with-url` metric ids, or I could imagine us maybe saying that in a future update we might add another type of `MetricFilter` that supports a regex value for the `:url` field.  Not sure.
6. I think it would be useful to have a `MetricFilter` that allowed me to ask for all of the `with-metric-id` metrics, without having to also specify a specific `metric-id`.
7. When I make two requests, one with a metric-id of ["puppetlabs" "jira"] and one with a metric-id of ["puppetlabs" "confluence"], and then I call `get-client-metrics-data` with a filter that includes `:metric-id ["puppetlabs"]`, I *only* get back the aggregate data.  I think I was kind of expecting that I'd get that *and* the values for the child nodes in that hierarchy.  I can imagine both use cases being desirable, so maybe we could add something to the `MetricFilter` that would allow you to flag whether you wanted only the exact match or everything underneath that in the tree.  It's possible that we already discussed this in our design meeting and agreed on the current approach, I can't remember.

Sorry, I know that's a lot of stuff :)  And I realize that the requirements for this have grown a bunch as we've been hacking away at it, so thanks for your patience.

I don't necessarily think that everything above needs to be addressed for this PR, just wanted to capture my thoughts.  It might be worth us setting up another design huddle meeting to prioritize these questions, decide which things we think are important, in what order, etc.  And figure out which of them would be breaking API changes if we added them at a later date.  I'd like to get to a point where we have a plan for the minimum set of changes we think are necessary to get this merged, and it's OK if we don't think it's quite in a releaseable state after the first PR is merged, but I think we need to try to minimize the amount of time that the repo is not in a releaseable state, so deciding on what exactly we think are the minimum requirements for an initial API and what things could be added later without breaking that API seems like the way to go.